File: python/mining.py
from collections import deque
from enum import Enum

from anvil import Region

class simpleBlock:
    """
    An abstraction of the Block class from the anvil library. This class consists of the coordinates of a block along with the name.
    """
    def __init__(self, coords: tuple[int, int, int], block: str):
        """
        Parameters
        ----------
        coords
            The coordinates that the block occupies
        block
            The name of the block ie. 'gold_ore'
        """
        self.x = coords[0]
        self.y = coords[1]
        self.z = coords[2]
        self.coords = coords
        self.block = block
    
    def __eq__(self, other) -> bool:
        """Only checks that the coordinates are the same"""
        return self.x == other.x and self.y == other.y and self.z == other.z

class Direction(Enum):
    """Self explanatory, only north, south, east, and west."""
    NORTH = 0,
    SOUTH = 1,
    EAST = 2,
    WEST = 3

def dirCoords(direction: Direction, coords: tuple[int, int, int], amount: int=1) -> tuple[int, int, int]:
    temp = list(coords)
    if direction == Direction.NORTH:
        temp[2] -= amount
    elif direction == Direction.SOUTH:
        temp[2] += amount
    elif direction == Direction.EAST:
        temp[0] += amount
    elif direction == Direction.WEST:
        temp[0] -= amount
    return tuple(temp)

def twoByOne(region: Region, direction: Direction, coords: tuple[int, int, int]) -> list[simpleBlock]:
    # [ ][x][ ]
    # [x][x][x]
    # [x][x][x]
    # [ ][x][ ]
    blocks = []
    xrange = [0]
    zrange = [0]
    if direction == Direction.NORTH or direction == Direction.SOUTH:
        xrange = range(-1, 2)
    if direction == Direction.EAST or direction == Direction.WEST:
        zrange = range(-1, 2)
    for x in xrange:
        for y in range(0, 2): # The player character is 2 blocks tall and the y coord represents the feet
            for z in zrange:
                blocks.append(simpleBlock(tuple((coords[0] + x, coords[1] + y, coords[2] + z)), getBlock(region, tuple((coords[0] + x, coords[1] + y, coords[2] + z)))))
    blocks.append(simpleBlock(tuple((coords[0], coords[1] - 1, coords[2])), getBlock(region, tuple((coords[0], coords[1] - 1, coords[2])))))
    blocks.append(simpleBlock(tuple((coords[0], coords[1] + 2, coords[2])), getBlock(region, tuple((coords[0], coords[1] + 2, coords[2])))))
    return blocks

# Maybe this should be made into a more generalized function
def oneByOne(region: Region, direction: Direction, coords: tuple((int, int, int))):
    # [ ][x][ ]
    # [x][x][x]
    # [ ][x][ ]
    blocks = []
    xrange = [0]
    zrange = [0]
    if direction == Direction.NORTH or direction == Direction.SOUTH:
        xrange = range(-1, 2)
    if direction == Direction.EAST or direction == Direction.WEST:
        zrange = range(-1, 2)
    for x in xrange:
        for y in range(0, 1):
            for z in zrange:
                blocks.append(simpleBlock(tuple((coords[0] + x, coords[1] + y, coords[2] + z)), getBlock(region, tuple((coords[0] + x, coords[1] + y, coords[2] + z)))))
    blocks.append(simpleBlock(tuple((coords[0], coords[1] - 1, coords[2])), getBlock(region, tuple((coords[0], coords[1] - 1, coords[2])))))
    blocks.append(simpleBlock(tuple((coords[0], coords[1] + 1, coords[2])), getBlock(region, tuple((coords[0], coords[1] + 1, coords[2])))))
    return blocks

def oneByOneEnd(region: Region, direction: Direction, coords: tuple((int, int, int))):
    blocks = oneByOne(region, direction, coords)
    if direction == Direction.NORTH:
        blocks.append(simpleBlock(tuple((coords[0], coords[1], coords[2] - 1)), getBlock(region, tuple((coords[0], coords[1], coords[2] - 1)))))
    elif direction == Direction.SOUTH:
        blocks.append(simpleBlock(tuple((coords[0], coords[1], coords[2] + 1)), getBlock(region, tuple((coords[0], coords[1], coords[2] + 1)))))
    elif direction == Direction.EAST:
        blocks.append(simpleBlock(tuple((coords[0] + 1, coords[1], coords[2])), getBlock(region, tuple((coords[0] + 1, coords[1], coords[2])))))
    elif direction == Direction.WEST:
        blocks.append(simpleBlock(tuple((coords[0] - 1, coords[1], coords[2])), getBlock(region, tuple((coords[0] - 1, coords[1], coords[2])))))
    return blocks

def twoByOneEnd(region: Region, direction: Direction, coords: tuple((int, int, int))):
    blocks = []
    if direction == Direction.NORTH:
        blocks.append(simpleBlock(tuple((coords[0], coords[1], coords[2] - 1)), getBlock(region, tuple((coords[0], coords[1], coords[2] - 1)))))
        blocks.append(simpleBlock(tuple((coords[0], coords[1] + 1, coords[2] - 1)), getBlock(region, tuple((coords[0], coords[1] + 1, coords[2] - 1)))))
    elif direction == Direction.SOUTH:
        blocks.append(simpleBlock(tuple((coords[0], coords[1], coords[2] + 1)), getBlock(region, tuple((coords[0], coords[1], coords[2] + 1)))))
        blocks.append(simpleBlock(tuple((coords[0], coords[1] + 1, coords[2] + 1)), getBlock(region, tuple((coords[0], coords[1] + 1, coords[2] + 1)))))
    elif direction == Direction.EAST:
        blocks.append(
            simpleBlock(
                tuple(
                    (coords[0] + 1, coords[1], coords[2])
                ), getBlock(
                    region, tuple(
                    (coords[0] + 1, coords[1], coords[2])
                    )
                )
            )
        )
        blocks.append(simpleBlock(tuple((coords[0] + 1, coords[1] + 1, coords[2])), getBlock(region, tuple((coords[0] + 1, coords[1] + 1, coords[2])))))
    elif direction == Direction.WEST:
        blocks.append(simpleBlock(tuple((coords[0] - 1, coords[1], coords[2])), getBlock(region, tuple((coords[0] - 1, coords[1], coords[2])))))
        blocks.append(simpleBlock(tuple((coords[0] - 1, coords[1] + 1, coords[2])), getBlock(region, tuple((coords[0] - 1, coords[1] + 1, coords[2])))))
    return blocks

# For the start of a poke the only unique block that is added from it is the block above the starting block
def pokeStart(region: Region, direction: Direction, coords: tuple((int, int, int))):
    blocks = []
    blocks.append(simpleBlock(tuple((coords[0], coords[1] + 1, coords[2])), getBlock(region, tuple((coords[0], coords[1] + 1, coords[2])))))
    return blocks

# queue has a item, add all of the surrounding blocks that are not already expanded and are part of the target list
# remove the first item from the list and add it to the expanded list
# repeat until there are no more items in the queue
# return the expanded list
def veinExpansionIterable(region: Region, valid: dict[str, str],targetx: int, targety: int, targetz: int):
    deq = deque()
    deq.append(simpleBlock(tuple((targetx, targety, targetz)), getBlock(region, tuple((targetx, targety, targetz)))))
    expanded = []
    current = 1

    while len(deq) > 0:
        block = deq.popleft()
        expanded.append(block)
        for x in range(-1, 2):
            for y in range(-1, 2):
                for z in range(-1, 2):
                    if not (x == 0 and y == 0 and z == 0):
                        adj = simpleBlock(tuple((block.x + x, block.y + y, block.z + z)), getBlock(region, tuple((block.x + x, block.y + y, block.z + z))))
                        if adj.block.id in valid:
                            found = False
                            for b in expanded:
                                if b.x == adj.x and b.y == adj.y and b.z == adj.z:
                                    found = True
                            for b in deq:
                                if b.x == adj.x and b.y == adj.y and b.z == adj.z:
                                    found = True
                            if not found:
                                deq.append(adj)
        current += 1
    
    return expanded



# This is a helper function to remove duplicates, this could likely be accomplished with a lambda function
def veinExpansionTrimming(found: list[simpleBlock]):
    sublist = []
    for i in range(len(found)):
        if found[i] not in sublist:
            sublist.append(found[i])
    return sublist

def getBlock(region: Region, coords: tuple((int, int, int))) -> str:
    """
    Gets a block id from a region file using input coordinates.

    Parameters
    ----------
    region
        The region file to be searched
    coords
        The coordinates of the block

    Returns
    -------
    The name block id of the block
    """
    chunkx = coords[0] // 16
    chunkz = coords[2] // 16
    chunk = region.get_chunk(chunkx, chunkz)
    return chunk.get_block(coords[0] % 16, coords[1], coords[2] % 16).id

File: src/mining.rs
use core::panic;
use std::{collections::{HashMap, VecDeque}, time::Instant};

use mvp_anvil::region::Region;

use crate::CachingRegion;

#[derive(Clone, PartialEq)]
pub enum Direction {
    North,
    South,
    East,
    West,
}

#[derive(Clone, Debug)]
pub struct SimpleBlock {
    pub block: String,
    pub x: i32,
    pub y: i32,
    pub z: i32,
}

impl PartialEq for SimpleBlock {
    fn eq(&self, other: &Self) -> bool {
        if self.x == other.x && self.y == other.y && self.z == other.z {
            return true;
        }
        return false;
    }
}

impl SimpleBlock {
    pub fn new(coords: (i32, i32, i32), block: String) -> SimpleBlock {
        return SimpleBlock {
            block,
            x: coords.0,
            y: coords.1,
            z: coords.2,
        };
    }

    pub fn get_coords(self) -> (i32, i32, i32) {
        return (self.x, self.y, self.z);
    }
}

pub fn shift_coords(
    direction: &Direction,
    mut coords: (i32, i32, i32),
    amount: i32,
) -> (i32, i32, i32) {
    match direction {
        Direction::North => coords.2 -= amount,
        Direction::South => coords.2 += amount,
        Direction::East => coords.0 += amount,
        Direction::West => coords.0 -= amount,
    };
    return coords;
}

/// Takes a region file and coordinates and returns the name of the block at that location. This handles the intermediary step of determining the x and z [chunks](`mvp_anvil::chunk::Chunk`) that correspond with the coordinates.
///
/// * `region` - The [region](`mvp_anvil::region::Region`) file that the block will be retrieved from.
/// * `coords` - The tuple of xyz coordinates of the block.
pub fn get_block(region: &mut CachingRegion, coords: (i32, i32, i32)) -> String {
    
    let chunk_x = coords.0 / 16;
    let chunk_z = coords.2 / 16;
    // let t2 = Instant::now();
    let chunk = region.get_chunk(chunk_x as usize, chunk_z as usize);
    // println!("Getting chunk took {} milliseconds", t2.elapsed().as_millis());
    let timer = Instant::now();
    let block = chunk.get_block(coords.0 % 16, coords.1, coords.2 % 16).id;
    // println!("Getting block took {} ns", timer.elapsed().as_nanos());
    return block;
}

fn two_by_one_single(
    region: &mut CachingRegion,
    direction: &Direction,
    coords: (i32, i32, i32),
) -> (Vec<SimpleBlock>, u32, u32) {
    let mut blocks = Vec::new();
    let mut x_range = 0..1;
    let mut z_range = 0..1;
    match direction {
        Direction::North | Direction::South => x_range = -1..2,
        Direction::East | Direction::West => z_range = -1..2,
    };
    for x in x_range {
        for y in 0..2 {
            for z in z_range.clone() {
                let new_coords = (coords.0 + x, coords.1 + y, coords.2 + z);
                blocks.push(SimpleBlock::new(new_coords, get_block(region, new_coords)));
            }
        }
    }
    blocks.push(SimpleBlock::new(
        (coords.0, coords.1 - 1, coords.2),
        get_block(region, (coords.0, coords.1 - 1, coords.2)),
    ));
    blocks.push(SimpleBlock::new(
        (coords.0, coords.1 + 2, coords.2),
        get_block(region, (coords.0, coords.1 + 2, coords.2)),
    ));
    return (blocks, 2, 8);
}

pub fn two_by_one_length(
    region: &mut CachingRegion,
    direction: &Direction,
    coords: (i32, i32, i32),
    length: i32,
) -> (Vec<SimpleBlock>, u32, u32) {
    let mut results = (Vec::new(), 0, 0);
    for n in 0..length {
        // let direction = direction.clone();
        // let direction_dup = direction.clone();
        let mut res = two_by_one_single(region, direction, shift_coords(direction, coords, n));
        results.0.append(&mut res.0);
        results.1 += res.1;
        results.2 += res.2;
    }
    return results;
}

pub fn two_by_one_end(
    region: &mut CachingRegion,
    direction: &Direction,
    coords: (i32, i32, i32),
) -> (Vec<SimpleBlock>, u32, u32) {
    let mut results = (Vec::new(), 0, 2);
    match direction {
        Direction::North => {
            results.0.push(SimpleBlock::new(
                (coords.0, coords.1, coords.2 - 1),
                get_block(region, (coords.0, coords.1, coords.2 - 1)),
            ));
            results.0.push(SimpleBlock::new(
                (coords.0, coords.1 + 1, coords.2 - 1),
                get_block(region, (coords.0, coords.1 + 1, coords.2 - 1)),
            ));
        }
        Direction::South => {
            results.0.push(SimpleBlock::new(
                (coords.0, coords.1, coords.2 + 1),
                get_block(region, (coords.0, coords.1, coords.2 + 1)),
            ));
            results.0.push(SimpleBlock::new(
                (coords.0, coords.1 + 1, coords.2 + 1),
                get_block(region, (coords.0, coords.1 + 1, coords.2 + 1)),
            ));
        }
        Direction::East => {
            results.0.push(SimpleBlock::new(
                (coords.0 + 1, coords.1, coords.2),
                get_block(region, (coords.0 + 1, coords.1, coords.2)),
            ));
            results.0.push(SimpleBlock::new(
                (coords.0 + 1, coords.1 + 1, coords.2),
                get_block(region, (coords.0 + 1, coords.1 + 1, coords.2)),
            ));
        }
        Direction::West => {
            results.0.push(SimpleBlock::new(
                (coords.0 - 1, coords.1, coords.2),
                get_block(region, (coords.0 - 1, coords.1, coords.2)),
            ));
            results.0.push(SimpleBlock::new(
                (coords.0 - 1, coords.1 + 1, coords.2),
                get_block(region, (coords.0 - 1, coords.1 + 1, coords.2)),
            ));
        }
    }
    return results;
}

fn one_by_one_single(
    region: &mut CachingRegion,
    direction: &Direction,
    coords: (i32, i32, i32),
) -> (Vec<SimpleBlock>, u32, u32) {
    let mut blocks = Vec::new();
    let mut x_range = 0..1;
    let mut z_range = 0..1;
    match direction {
        Direction::North | Direction::South => x_range = -1..2,
        Direction::East | Direction::West => z_range = -1..2,
    };
    for x in x_range {
        for y in 0..1 {
            let z_range = z_range.clone();
            for z in z_range {
                let new_coords = (coords.0 + x, coords.1 + y, coords.2 + z);
                blocks.push(SimpleBlock::new(new_coords, get_block(region, new_coords)));
            }
        }
    }
    blocks.push(SimpleBlock::new(
        (coords.0, coords.1 - 1, coords.2),
        get_block(region, (coords.0, coords.1 - 1, coords.2)),
    ));
    blocks.push(SimpleBlock::new(
        (coords.0, coords.1 + 1, coords.2),
        get_block(region, (coords.0, coords.1 + 2, coords.2)),
    ));
    return (blocks, 1, 5);
}

fn one_by_one_length(
    region: &mut CachingRegion,
    direction: &Direction,
    coords: (i32, i32, i32),
    length: i32,
) -> (Vec<SimpleBlock>, u32, u32) {
    let mut results = (Vec::new(), 0, 0);
    for n in 0..length {
        let mut res = one_by_one_single(region, direction, shift_coords(direction, coords, n));
        results.0.append(&mut res.0);
        results.1 += res.1;
        results.2 += res.2;
    }
    return results;
}

fn one_by_one_end(
    region: &mut CachingRegion,
    direction: &Direction,
    coords: (i32, i32, i32),
) -> (Vec<SimpleBlock>, u32, u32) {
    let mut results = (Vec::new(), 0, 1);
    match direction {
        Direction::North => {
            results.0.push(SimpleBlock::new(
                (coords.0, coords.1, coords.2 - 1),
                get_block(region, (coords.0, coords.1, coords.2 - 1)),
            ));
        }
        Direction::South => {
            results.0.push(SimpleBlock::new(
                (coords.0, coords.1, coords.2 + 1),
                get_block(region, (coords.0, coords.1, coords.2 + 1)),
            ));
        }
        Direction::East => {
            results.0.push(SimpleBlock::new(
                (coords.0 + 1, coords.1, coords.2),
                get_block(region, (coords.0 + 1, coords.1, coords.2)),
            ));
        }
        Direction::West => {
            results.0.push(SimpleBlock::new(
                (coords.0 - 1, coords.1, coords.2),
                get_block(region, (coords.0 - 1, coords.1, coords.2)),
            ));
        }
    }
    return results;
}

fn poke_start(region: &mut CachingRegion, coords: (i32, i32, i32)) -> (Vec<SimpleBlock>, u32, u32) {
    let mut results = (Vec::new(), 0, 1);
    results.0.push(SimpleBlock::new(
        (coords.0, coords.1 + 1, coords.2),
        get_block(region, (coords.0, coords.1 + 1, coords.2)),
    ));
    return results;
}

pub fn poke(
    region: &mut CachingRegion,
    direction: &Direction,
    coords: (i32, i32, i32),
    depth: i32,
) -> (Vec<SimpleBlock>, u32, u32) {
    if depth < 1 {
        panic!("Poke should be at least 1 block in depth")
    }
    let mut results = (Vec::new(), 0, 0);
    let mut res = poke_start(region, coords);
    results.0.append(&mut res.0);
    results.1 += res.1;
    results.2 += res.2;
    let mut res = one_by_one_length(
        region,
        direction,
        shift_coords(direction, coords, 1),
        depth - 1,
    );
    results.0.append(&mut res.0);
    results.1 += res.1;
    results.2 += res.2;
    let mut res = one_by_one_end(
        region,
        direction,
        shift_coords(direction, coords, depth - 1),
    );
    results.0.append(&mut res.0);
    results.1 += res.1;
    results.2 += res.2;
    return results;
}
